sdk/core/Azure.Core/src/Shared/MessagingClientDiagnostics.cs

```diff
@@ -0,0 +1,136 @@
+            _scopeFactory = new DiagnosticScopeFactory(clientNamespace, resourceProviderNamespace, true, false);
+        }
+
+        public DiagnosticScope CreateScope(
+            string activityName,
+            DiagnosticScope.ActivityKind kind,
+            MessagingDiagnosticOperation diagnosticOperation)
+        {
+            DiagnosticScope scope = _scopeFactory.CreateScope(activityName, kind);
+            if (ActivityExtensions.SupportsActivitySource())
+            {
+                scope.AddAttribute(MessagingSystem, _messagingSystem);
+                scope.AddAttribute(MessagingOperation, diagnosticOperation.ToString());
+                scope.AddAttribute(NetPeerName, _fullyQualifiedNamespace);
+                scope.AddAttribute(diagnosticOperation == MessagingDiagnosticOperation.Publish ? DestinationName : SourceName, _entityPath);
+            }
+            else
+            {
+                scope.AddAttribute(Component, _messagingSystem);
+                scope.AddAttribute(PeerAddress, _fullyQualifiedNamespace);
+                scope.AddAttribute(MessageBusDestination, _entityPath);
+            }
+
+            return scope;
+        }
+
+        /// <summary>
+        ///   Attempts to extract a diagnostic id from a message's properties.
+        /// </summary>
+        ///
+        /// <param name="properties">The properties holding the diagnostic id.</param>
+        /// <param name="id">The value of the diagnostics identifier assigned to the event. </param>
+        ///
+        /// <returns><c>true</c> if the message properties contained the diagnostic id; otherwise, <c>false</c>.</returns>
+        public static bool TryExtractDiagnosticId(IReadOnlyDictionary<string, object> properties, out string? id)
+        {
+            id = null;
+
+            if (properties.TryGetValue(DiagnosticIdAttribute, out var objectId) && objectId is string stringId)
+            {
+                id = stringId;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        ///   Attempts to extract a diagnostic id from a message's properties.
+        /// </summary>
+        ///
+        /// <param name="properties">The properties holding the diagnostic id.</param>
+        /// <param name="id">The value of the diagnostics identifier assigned to the event. </param>
+        ///
+        /// <returns><c>true</c> if the message properties contained the diagnostic id; otherwise, <c>false</c>.</returns>
+        public static bool TryExtractDiagnosticId(IDictionary<string, object> properties, out string? id)
+        {
+            id = null;
+
+            if (properties.TryGetValue(DiagnosticIdAttribute, out var objectId) && objectId is string stringId)
+            {
+                id = stringId;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Instrument the message properties for tracing.
```

Comment:
Should we consider adding a note here that this does modify the message properties and increase the size?